### DISCUSSION TITLE
What's the optimal hardware config for EMR?

### DISCUSSION THREAD

**Discussion by shy00n** [moj-analytical-services/splink#194] [Q&A]
I reworked the code below to follow the example & the example took 30 min + before I killed it.
Using below. Why is the simple example take so long? Can you suggest a better config for the hw?
I am trying to match 8M vs 250M dataframes. 
Thanks much in adv,

ig-2BZ6QQQRGWB5J | Running | MASTERMaster - 1 | m5d.2xlarge8 vCore, 32 GiB memory, 300 SSD GB storageEBS Storage:64 GiB | 1Instances | On-demand
-- | -- | -- | -- | -- | --
ig-O14JGC2HXF4G | Running | CORECore - 2 | m5.2xlarge8 vCore, 32 GiB memory, EBS only storageEBS Storage:128 GiB | 2InstancesResize | On-demand

(4 instances on demand)
Configuration details
Release label:emr-6.2.0
Hadoop distribution:Amazon
Applications:Spark 3.0.1

Application user interfaces
Persistent user interfaces:Spark history server, YARN timeline server
On-cluster user interfaces:Not EnabledEnable an SSH Connection
Network and hardware
Availability zone:us-east-1a
Master:Running1m5d.2xlarge
Core:Running2m5.2xlarge
Task:--
Cluster scaling:EMR-managed scaling

```
import logging
from splink import Splink
from athena_ojdbc.spark_hlper import (
    f,
    token_nulls_XXX,
    token_null_counts,
    null_counts,
    get_spark,
    read_s3_parquet,
    read_foldr,
    write_parquet_ymd,
    add_subseq_in_grp,
    hash_a_columns,
    StringType,
    DoubleType,
)
from pii_hlper import cor_tokens_match, df4_cols, df10_cols

spark = get_spark("splink_run", "15g")

spark.udf.registerJavaFunction(
    "jaro_winkler_sim",
    "uk.gov.moj.dash.linkage.JaroWinklerSimilarity",
    DoubleType(),
)

spark.udf.registerJavaFunction(
    "Dmetaphone", "uk.gov.moj.dash.linkage.DoubleMetaphone", StringType()
)

WRITE_PARTITIONS = ["token"]
SANDBOX_PATH = "s3://sandboxes-analysis/hoon/"

df_1 = spark.read.parquet(SANDBOX_PATH + "splink_data/fake_df_l.parquet")
df_1 = df_1.withColumn("source_dataset", f.lit("df_1"))
df_2 = spark.read.parquet(SANDBOX_PATH + "splink_data/fake_df_r.parquet")
df_2 = df_2.withColumn("source_dataset", f.lit("df_2"))
print(f"The count of rows in `df_1` is {df_1.count()}")
df_1.show(5)
print(f"The count of rows in `df_2` is {df_2.count()}")
df_2.show(5)


sql_case_expression = """
CASE
WHEN first_name_l = first_name_r AND surname_l = surname_r THEN 4
WHEN first_name_l = surname_r AND surname_l = first_name_r THEN 3
WHEN first_name_l = first_name_r THEN 2
WHEN surname_l = surname_r THEN 1
ELSE 0
END
"""

settings = {
    "link_type": "link_only",
    "max_iterations": 20,
    "blocking_rules": [],
    "comparison_columns": [
        {
            "custom_name": "name_inversion",
            "custom_columns_used": ["first_name", "surname"],
            "case_expression": sql_case_expression,
            "num_levels": 5,
        },
        {"col_name": "city", "num_levels": 3},
        {"col_name": "email", "num_levels": 3},
        {"col_name": "dob"},
    ],
    "additional_columns_to_retain": ["group"],
}


linker = Splink(settings, [df_1, df_2], spark)
df_e = linker.get_scored_comparisons()

# Later, we will make term frequency adjustments.
# Persist caches these results in memory, preventing them having to be recomputed when we make these adjustments.
df_e.persist()
```

**Comment by RobinL** [comment]
The hardware looks fine.  Running against the demo data you're using (819 and 181 rows) with no blocking rules results in 148,239 comparisons (comparing every record with every other).  I would expect this to run fine on a single low specced machine , with a runtime of perhaps a minutes or so.

Did it print any errors or logs in that time?  Are you sure the code ran successfully up to the `df_e = linker.get_scored_comparisons()` line?


For what it's worth, we use AWS Glue, which is built on top of EMR.  We use the standard worker type - details [here](https://docs.aws.amazon.com/glue/latest/dg/add-job.html).  Depending on the job size, we may use a cluster side of between 4 and 15 of these.  4 would typically be adequate for datasets of <1 million records with blocking rules being used.  

Your job of 8M vs 250M will probably require quite a large cluster.  I'd start with quite restrictive blocking rules to make sure the number of comparisons generated stays fairly small (say, less than 1 billion rows).
